File: protocol.py
"""
protocol.py - minimal Modbus/TCP framing for VoltGuard

Real Modbus/TCP ADU layout:
  MBAP header (7 bytes): transaction_id(2) protocol_id(2) length(2) unit_id(1)
  PDU: function_code(1) + payload

We only implement function code 0x06 (Write Single Register), since that's
the one used to command pump RPM in the mock pipeline. Good enough for
week 1 - swap in pymodbus or a real Scapy dissector later if you need the
full function code set.
"""

import struct
import random

FUNC_WRITE_SINGLE_REGISTER = 0x06
PUMP_RPM_REGISTER = 0x0001  # the register address that represents pump RPM

_txn_counter = 0


def _next_txn_id():
    global _txn_counter
    _txn_counter = (_txn_counter + 1) % 65535
    return _txn_counter


def build_write_register_command(rpm, unit_id=1):
    """Build a raw Modbus/TCP frame commanding the pump to spin at `rpm`."""
    if not (0 <= rpm <= 65535):
        raise ValueError("rpm must fit in a 16-bit register (0-65535)")

    txn_id = _next_txn_id()
    protocol_id = 0
    pdu = struct.pack(">BHH", FUNC_WRITE_SINGLE_REGISTER, PUMP_RPM_REGISTER, rpm)
    length = len(pdu) + 1  # +1 for unit_id
    mbap = struct.pack(">HHHB", txn_id, protocol_id, length, unit_id)
    return mbap + pdu


def parse_frame(raw):
    """Parse a raw Modbus/TCP frame back into its fields. Raises ValueError on garbage."""
    if len(raw) < 8:
        raise ValueError("frame too short to be a valid Modbus/TCP ADU")

    txn_id, protocol_id, length, unit_id = struct.unpack(">HHHB", raw[:7])
    pdu = raw[7:]

    if protocol_id != 0:
        raise ValueError(f"unexpected protocol_id {protocol_id}, expected 0")
    if len(pdu) != length - 1:
        raise ValueError("length field doesn't match actual payload size")

    func_code = pdu[0]
    if func_code != FUNC_WRITE_SINGLE_REGISTER:
        raise ValueError(f"unsupported function code 0x{func_code:02x}")

    reg_addr, reg_value = struct.unpack(">HH", pdu[1:5])

    return {
        "transaction_id": txn_id,
        "unit_id": unit_id,
        "function_code": func_code,
        "register_addr": reg_addr,
        "rpm": reg_value,
    }


EXCEPTION_BIT = 0x80
EXC_SERVER_DEVICE_FAILURE = 0x04  # what we return when VoltGuard drops a command


def build_success_response(txn_id, unit_id, reg_addr, reg_value):
    """A real Modbus write-single-register success reply just echoes the
    request's address and value back - that's the wire-level 'ack'."""
    pdu = struct.pack(">BHH", FUNC_WRITE_SINGLE_REGISTER, reg_addr, reg_value)
    length = len(pdu) + 1
    mbap = struct.pack(">HHHB", txn_id, 0, length, unit_id)
    return mbap + pdu


def build_exception_response(txn_id, unit_id, exception_code=EXC_SERVER_DEVICE_FAILURE):
    """What VoltGuard sends back instead of forwarding a dangerous command -
    a real Modbus exception response, function code with the high bit set."""
    pdu = struct.pack(">BB", FUNC_WRITE_SINGLE_REGISTER | EXCEPTION_BIT, exception_code)
    length = len(pdu) + 1
    mbap = struct.pack(">HHHB", txn_id, 0, length, unit_id)
    return mbap + pdu


def parse_response(raw):
    """Parse whatever comes back from the gateway/PLC - success echo or exception."""
    txn_id, protocol_id, length, unit_id = struct.unpack(">HHHB", raw[:7])
    pdu = raw[7:]
    func_code = pdu[0]
    if func_code & EXCEPTION_BIT:
        return {"ok": False, "transaction_id": txn_id, "exception_code": pdu[1]}
    reg_addr, reg_value = struct.unpack(">HH", pdu[1:5])
    return {"ok": True, "transaction_id": txn_id, "register_addr": reg_addr, "rpm": reg_value}


def generate_normal_command():
    """A command a real plant operator would actually send."""
    rpm = random.randint(500, 2800)
    return build_write_register_command(rpm)


def generate_malicious_command():
    """A syntactically valid but physically dangerous command - the kind an
    IT firewall waves straight through because nothing about the packet
    itself is malformed."""
    rpm = random.choice([
        random.randint(15000, 30000),
        50000,  # the exact scenario from the problem statement
        random.randint(40000, 65535),
    ])
    return build_write_register_command(rpm)


if __name__ == "__main__":
    normal = generate_normal_command()
    malicious = generate_malicious_command()
    print("normal frame:", normal.hex(), "->", parse_frame(normal))
    print("malicious frame:", malicious.hex(), "->", parse_frame(malicious))

File: protocol.rs
//! protocol.rs - Modbus/TCP framing, byte-for-byte identical to protocol.py
//! and modbus_parser.cpp. Three languages, one wire format - that's the
//! actual proof this is a real protocol, not three incompatible toys.

pub const FUNC_WRITE_SINGLE_REGISTER: u8 = 0x06;
pub const EXCEPTION_BIT: u8 = 0x80;
pub const EXC_SERVER_DEVICE_FAILURE: u8 = 0x04;
pub const PUMP_RPM_REGISTER: u16 = 0x0001;

#[derive(Debug, Clone, Copy)]
pub struct ParsedFrame {
    pub transaction_id: u16,
    pub unit_id: u8,
    pub function_code: u8,
    pub register_addr: u16,
    pub rpm: u16,
}

#[allow(dead_code)] // used by tests; kept public for a future Rust traffic-generator utility
pub fn build_write_register_command(txn_id: u16, unit_id: u8, rpm: u16) -> Vec<u8> {
    let mut pdu = Vec::with_capacity(5);
    pdu.push(FUNC_WRITE_SINGLE_REGISTER);
    pdu.extend_from_slice(&PUMP_RPM_REGISTER.to_be_bytes());
    pdu.extend_from_slice(&rpm.to_be_bytes());

    let length: u16 = (pdu.len() + 1) as u16; // +1 for unit_id

    let mut frame = Vec::with_capacity(7 + pdu.len());
    frame.extend_from_slice(&txn_id.to_be_bytes());
    frame.extend_from_slice(&0u16.to_be_bytes()); // protocol_id
    frame.extend_from_slice(&length.to_be_bytes());
    frame.push(unit_id);
    frame.extend_from_slice(&pdu);
    frame
}

#[allow(dead_code)] // kept for parity/completeness with protocol.py and modbus_parser.cpp,
                     // even though this gateway currently relays the real PLC's own reply
pub fn build_success_response(txn_id: u16, unit_id: u8, reg_addr: u16, reg_value: u16) -> Vec<u8> {
    let mut pdu = Vec::with_capacity(5);
    pdu.push(FUNC_WRITE_SINGLE_REGISTER);
    pdu.extend_from_slice(&reg_addr.to_be_bytes());
    pdu.extend_from_slice(&reg_value.to_be_bytes());
    let length = (pdu.len() + 1) as u16;

    let mut frame = Vec::with_capacity(7 + pdu.len());
    frame.extend_from_slice(&txn_id.to_be_bytes());
    frame.extend_from_slice(&0u16.to_be_bytes());
    frame.extend_from_slice(&length.to_be_bytes());
    frame.push(unit_id);
    frame.extend_from_slice(&pdu);
    frame
}

pub fn build_exception_response(txn_id: u16, unit_id: u8) -> Vec<u8> {
    let pdu = vec![FUNC_WRITE_SINGLE_REGISTER | EXCEPTION_BIT, EXC_SERVER_DEVICE_FAILURE];
    let length = (pdu.len() + 1) as u16;

    let mut frame = Vec::with_capacity(7 + pdu.len());
    frame.extend_from_slice(&txn_id.to_be_bytes());
    frame.extend_from_slice(&0u16.to_be_bytes());
    frame.extend_from_slice(&length.to_be_bytes());
    frame.push(unit_id);
    frame.extend_from_slice(&pdu);
    frame
}

#[derive(Debug)]
pub enum ParseError {
    TooShort,
    BadProtocolId(u16),
    LengthMismatch,
    UnsupportedFunctionCode(u8),
}

impl std::fmt::Display for ParseError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            ParseError::TooShort => write!(f, "frame too short to be a valid Modbus/TCP ADU"),
            ParseError::BadProtocolId(p) => write!(f, "unexpected protocol_id {p}, expected 0"),
            ParseError::LengthMismatch => write!(f, "length field doesn't match actual payload size"),
            ParseError::UnsupportedFunctionCode(fc) => write!(f, "unsupported function code 0x{fc:02x}"),
        }
    }
}

pub fn parse_frame(raw: &[u8]) -> Result<ParsedFrame, ParseError> {
    if raw.len() < 8 {
        return Err(ParseError::TooShort);
    }

    let transaction_id = u16::from_be_bytes([raw[0], raw[1]]);
    let protocol_id = u16::from_be_bytes([raw[2], raw[3]]);
    let length = u16::from_be_bytes([raw[4], raw[5]]);
    let unit_id = raw[6];

    if protocol_id != 0 {
        return Err(ParseError::BadProtocolId(protocol_id));
    }

    let pdu_size = raw.len() - 7;
    if pdu_size != (length as usize).wrapping_sub(1) {
        return Err(ParseError::LengthMismatch);
    }

    let function_code = raw[7];
    if function_code != FUNC_WRITE_SINGLE_REGISTER {
        return Err(ParseError::UnsupportedFunctionCode(function_code));
    }

    let register_addr = u16::from_be_bytes([raw[8], raw[9]]);
    let rpm = u16::from_be_bytes([raw[10], raw[11]]);

    Ok(ParsedFrame { transaction_id, unit_id, function_code, register_addr, rpm })
}

#[derive(Debug)]
pub enum ResponseOutcome {
    Ok { rpm: u16, actual_pressure: Option<f64> },
    Exception { exception_code: u8 },
}

/// Parses whatever comes back from the PLC - success echo (optionally with
/// actual-pressure telemetry appended) or a Modbus exception response.
pub fn parse_response(raw: &[u8]) -> Result<ResponseOutcome, ParseError> {
    if raw.len() < 8 {
        return Err(ParseError::TooShort);
    }
    let pdu = &raw[7..];
    let func_code = pdu[0];

    if func_code & EXCEPTION_BIT != 0 {
        return Ok(ResponseOutcome::Exception { exception_code: pdu[1] });
    }

    let rpm = u16::from_be_bytes([pdu[3], pdu[4]]);
    let actual_pressure = if pdu.len() >= 7 {
        let actual_x10 = u16::from_be_bytes([pdu[5], pdu[6]]);
        Some(actual_x10 as f64 / 10.0)
    } else {
        None
    };
    Ok(ResponseOutcome::Ok { rpm, actual_pressure })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip_matches_python_frame() {
        // this exact hex was produced by protocol.py for txn_id=1, unit_id=1,
        // register_addr=1, rpm=2800 - same parity check used for the C++ parser
        let known_good = hex_decode("000100000006010600010af0");
        let parsed = parse_frame(&known_good).expect("should parse");
        assert_eq!(parsed.transaction_id, 1);
        assert_eq!(parsed.unit_id, 1);
        assert_eq!(parsed.register_addr, 1);
        assert_eq!(parsed.rpm, 2800);
    }

    #[test]
    fn build_then_parse_round_trip() {
        let raw = build_write_register_command(42, 1, 50000);
        let parsed = parse_frame(&raw).expect("should parse");
        assert_eq!(parsed.transaction_id, 42);
        assert_eq!(parsed.rpm, 50000);
    }

    #[test]
    fn rejects_short_frame() {
        assert!(matches!(parse_frame(&[0x00, 0x01, 0x02]), Err(ParseError::TooShort)));
    }

    fn hex_decode(s: &str) -> Vec<u8> {
        (0..s.len()).step_by(2).map(|i| u8::from_str_radix(&s[i..i + 2], 16).unwrap()).collect()
    }
}
